### ISSUE TITLE
Reseting max_agents 

### ISSUE THREAD

**Issue by anna-braun** [JuPedSim/jpscore#412]
In _traffic_constraints_ we can give doors the parameter _max_agents_. If _max_agents_ have passed a door it will be closed permanently. For several reasons it should be possible to reopen such doors and reset the counter of _max_agents_ when the doors are reopened. Here are two ideas for it:     
1.  Adapting the schedules so that they can be controlled by time and by max_agents.      
-> The doors will be closed after a defined time or when a maximum number of agents have passed the doors. The counter for the _max_agents_ should be reseted when the door reopens after the closing time:
```xml 
<time group_id="0" closing_time="20" max_agents="300">
    <t t="30"/>
    <t t="80"/>
</time>
```
2. Reset the counter of max_agents if the door is opened by an event.      
-> Set the counter of _max_agents_ in _traffic_constraints_ to zero when the door is opened by an event:

**Comment by chraibi** [comment]
I guess solution 2 is more intuitive. 

The original purpose of `max_agents` was to assure a (*permanent*) limit for the door. What if you simply take a very big value (inf)?

**Comment by schroedtert** [comment]
@chraibi Is the regulation of the flow of normal still working? For a simple case in the current commit 6b9e0f0b2d4469ecdcc326937e0d7f03fdf5cced it did not work for me. The is not closed after max_agents passed through it.

[412-reset-door.zip](https://github.com/JuPedSim/jpscore/files/3412113/412-reset-door.zip)

**Comment by chraibi** [comment]
It works fine, but, you use `dn=0`. 

However, according to http://www.jupedsim.org/jpscore/2016-11-01-inifile.html#traffic-constraints

> `dn:` number of agents to pass the door before triggering the process of flow regulation.

In [simulation.cpp](https://github.com/JuPedSim/jpscore/blob/develop/Simulation.cpp#L1112)
```
                 if( trans->GetPartialDoorUsage() ==  trans->GetDN() ) {
                      trans->regulateFlow(Pedestrian::GetGlobalTime());
                      trans->ResetPartialDoorUsage();
                 }
```
the condition is never triggered when `dn==0`.

Another *hack*: 

This all works only when you activate 

Because as we can see in this  [line](https://github.com/JuPedSim/jpscore/blob/develop/Simulation.cpp#L1069)

`Simulation::UpdateFlowAtDoors` is called when this conditio is true
```
    if (_config->ShowStatistics())
```

To summarize:  It works with

```
 <show_statistics>true</show_statistics> 
```

and 

```
 <door trans_id="0" caption="roomdoor0" state="open" dn="2" max_agents="5" outflow="2"/> 
```

@anna-braun: would it be enough to set the door's state to `tempClose` instead of `close`?

**Comment by anna-braun** [comment]
@chraibi (@anna-braun: would it be enough to set the door's state to tempClose instead of close?)
I guess temp close is also okay, if a reset is possible with this.

**Comment by schroedtert** [comment]
> @chraibi (@anna-braun: would it be enough to set the door's state to tempClose instead of close?)
> I guess temp close is also okay, if a reset is possible with this.

Temp close would have the effect, that peds would wait in front of that door instead of searching for a available exit. To our use case with train doors, I think close is more appropriate.

**Comment by anna-braun** [comment]
@schroedtert "close" is okay, if the agents are waiting at any place inside the room if there is no exit open.

**Comment by schroedtert** [comment]
At the moment agents which are within a room with no open transition will be deleted as they have no way to the outside. This will need some larger adjustments.

@anna-braun @chraibi One question: Do we want a particular event for resetting the counter of the door usage or should it be done when opening a door by event? I assume with introducing a new event we may get more control.

**Comment by anna-braun** [comment]
@schroedtert yes a new event could be good.
How did you think this could look like?

**Comment by schroedtert** [comment]
@anna-braun Have you tested it since #441 is merged? The new event is called `reset`. And can be used in events and shedules

**Comment by anna-braun** [comment]
@schroedtert 
Did we talk about it? For the newest version I get a segmentation fault by trying to reset the max_agents.
[20191029.zip](https://github.com/JuPedSim/jpscore/files/3783575/20191029.zip)

**Comment by schroedtert** [comment]
> @schroedtert
> Did we talk about it? For the newest version I get a segmentation fault by trying to reset the max_agents.
> [20191029.zip](https://github.com/JuPedSim/jpscore/files/3783575/20191029.zip)

@anna-braun When you set the doors at `t=0` to `state=temp_close` instead of `state=close` it works. With close you mark doors at not passable anymore during the simulation. We have to discuss and check whether all router work with all events correctly.

**Comment by anna-braun** [comment]
> @anna-braun When you set the doors at `t=0` to `state=temp_close` instead of `state=close` it works. With close you mark doors at not passable anymore during the simulation. We have to discuss and check whether all router work with all events correctly.

@schroedtert I tried temp_close as well, with the same result. But with the fix of @Ozaq it works now (using temp_close). But the agents do still not behave like I expect it. When a door is closed by max_agents, the agents should look for an other door, or not? And if there is none, they should wait. But we can talk about this, next week.
[traj-max_agents.txt](https://github.com/JuPedSim/jpscore/files/3793310/traj-max_agents.txt)

**Comment by schroedtert** [comment]
@anna-braun we should definitely talk about this. I think some things regarding the doors and close vs temp_close should be clarified.

**Comment by anna-braun** [comment]
The reset function works, But the door status of max_agents should be changed, see issue #533

**Comment by anna-braun** [comment]
@schroedtert 
There is one little thing, which should be changed again.
When the counter of max_agents is reseted, the cumulative counter in the door statistics is also set to zero. This should not be done, the statistics counter should go on.
[flow_exit_id_0_traj-max_agents.txt](https://github.com/JuPedSim/jpscore/files/3824313/flow_exit_id_0_traj-max_agents.txt)

**Comment by schroedtert** [comment]
@anna-braun could you please check if #543 works for you?

**Comment by anna-braun** [comment]
> @anna-braun could you please check if #543 works for you?

@schroedtert It works, thank you!
